tools/database/src/resharding.rs

```diff
@@ -0,0 +1,188 @@
+        tracing::info!(target: "resharding", ?state_roots, "state roots");
+
+        chain.custom_build_state_for_split_shards_postprocessing(
+            shard_uid,
+            &sync_hash,
+            state_roots,
+        )?;
+
+        if self.restore {
+            // In restore mode print database write statistics.
+            chain.runtime_adapter.store().get_store_statistics().map(|stats| {
+                stats.data.iter().for_each(|(metric, values)| {
+                    tracing::info!(target: "resharding", %metric, ?values);
+                })
+            });
+        }
+
+        Ok(())
+    }
+
+    fn get_store(&self, home_dir: &Path, config: &mut NearConfig) -> Result<Store, anyhow::Error> {
+        // Open hot and cold as usual.
+        let storage = open_storage(home_dir, config)?;
+
+        let cold_db = storage
+            .cold_db()
+            .expect("resharding tool can be used only on archival nodes")
+            .to_owned();
+        let hot_db = storage.into_inner(Temperature::Hot);
+
+        // We need real split db so that it correctly handles reads of missing
+        // values in the columns that are not in the cold db.
+        let split_db = SplitDB::new(hot_db, cold_db.clone());
+
+        let store = if self.restore {
+            // In 'restore' mode all changes are written directly into the recovery DB built upon the cold DB.
+            let recovery_db = Arc::new(RecoveryDB::new(cold_db));
+
+            // Prepare the full mixed db.
+            // It will read, in order, from hot, cold and recovery.
+            // It will write only to the recovery db.
+            let mixed_db = MixedDB::new(split_db, recovery_db, ReadOrder::ReadDBFirst);
+
+            // The only way to create a Store is to go through NodeStorage.
+            let storage = NodeStorage::new(mixed_db);
+            storage.get_hot_store()
+        } else {
+            let write_path =
+                self.write_path.as_ref().expect("write path must be set when not in recovery mode");
+
+            // Open write db.
+            let write_path = if write_path.is_absolute() {
+                PathBuf::from(&write_path)
+            } else {
+                home_dir.join(&write_path)
+            };
+            let write_path = write_path.as_path();
+            let write_config = &config.config.store;
+            let write_db =
+                RocksDB::open(write_path, write_config, Mode::ReadWrite, Temperature::Hot)?;
+            let write_db = Arc::new(write_db);
+
+            // Prepare the full mixed db.
+            // It will read, in order, from write, hot and cold.
+            // It will write only to the write db.
+            let mixed_db = MixedDB::new(split_db, write_db, ReadOrder::WriteDBFirst);
+
+            // The only way to create a Store is to go through NodeStorage.
+            let storage = NodeStorage::new(mixed_db);
+            storage.get_hot_store()
```

Comment:
Unless you are planning on using it I would get rid of this option to keep things clean and simple for now.

Reply:
I've used this option in the past and I will use it again to debug a failure.. so I'll probably keep it!